In [41]:
import pandas as pd
import numpy as np
import os

#to ignore warnings
import warnings
warnings.filterwarnings('ignore')

files=os.listdir('/Users/kanis/Documents/Sales_Data')
files

['Sales_April_2019.csv',
 'Sales_August_2019.csv',
 'Sales_December_2019.csv',
 'Sales_February_2019.csv',
 'Sales_January_2019.csv',
 'Sales_July_2019.csv',
 'Sales_June_2019.csv',
 'Sales_March_2019.csv',
 'Sales_May_2019.csv',
 'Sales_November_2019.csv',
 'Sales_October_2019.csv',
 'Sales_September_2019.csv']

In [42]:
#Read the files
df = pd.DataFrame()

for i in files:
    data=pd.read_csv(os.path.join('/Users/kanis/Documents/Sales_Data',i))
    df=pd.concat([df,data])

df

Order ID                     Product Quantity Ordered Price Each  \
0       176558        USB-C Charging Cable                2      11.95   
1          NaN                         NaN              NaN        NaN   
2       176559  Bose SoundSport Headphones                1      99.99   
3       176560                Google Phone                1        600   
4       176560            Wired Headphones                1      11.99   
...        ...                         ...              ...        ...   
11681   259353      AAA Batteries (4-pack)                3       2.99   
11682   259354                      iPhone                1        700   
11683   259355                      iPhone                1        700   
11684   259356      34in Ultrawide Monitor                1     379.99   
11685   259357        USB-C Charging Cable                1      11.95   

           Order Date                         Purchase Address  
0      04/19/19 08:46             917 1st St, Dallas, TX 75001  
1                 NaN                                      NaN  
2      04/07/19 22:30        682 Chestnut St, Boston, MA 02215  
3      04/12/19 14:38     669 Spruce St, Los Angeles, CA 90001  
4      04/12/19 14:38     669 Spruce St, Los Angeles, CA 90001  
...               ...                                      ...  
11681  09/17/19 20:56   840 Highland St, Los Angeles, CA 90001  
11682  09/01/19 16:00  216 Dogwood St, San Francisco, CA 94016  
11683  09/23/19 07:39     220 12th St, San Francisco, CA 94016  
11684  09/19/19 17:30   511 Forest St, San Francisco, CA 94016  
11685  09/30/19 00:18   250 Meadow St, San Francisco, CA 94016  

[186850 rows x 6 columns]

In [43]:
df.columns

Index(['Order ID', 'Product', 'Quantity Ordered', 'Price Each', 'Order Date',
       'Purchase Address'],
      dtype='object')

In [44]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 186850 entries, 0 to 11685
Data columns (total 6 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   Order ID          186305 non-null  object
 1   Product           186305 non-null  object
 2   Quantity Ordered  186305 non-null  object
 3   Price Each        186305 non-null  object
 4   Order Date        186305 non-null  object
 5   Purchase Address  186305 non-null  object
dtypes: object(6)
memory usage: 10.0+ MB


In [45]:
df.isnull().sum()

Order ID            545
Product             545
Quantity Ordered    545
Price Each          545
Order Date          545
Purchase Address    545
dtype: int64

In [46]:
#as null values are same and it means there exist a whole null row, then we drop all null values
df.dropna(inplace=True)
df.shape

(186305, 6)

#### Changing Data Type

In [47]:
#getting unique value for each category
df[['Order ID','Product','Quantity Ordered','Price Each','Order Date','Purchase Address']].describe()

Order ID               Product Quantity Ordered Price Each  \
count     186305                186305           186305     186305   
unique    178438                    20               10         24   
top     Order ID  USB-C Charging Cable                1      11.95   
freq         355                 21903           168552      21903   

        Order Date  Purchase Address  
count       186305            186305  
unique      142396            140788  
top     Order Date  Purchase Address  
freq           355               355

In [48]:
df['Quantity Ordered'].unique()

array(['2', '1', '3', '5', 'Quantity Ordered', '4', '7', '6', '8', '9'],
      dtype=object)

In [49]:
#create the filter to drop the text value from quantity ordered
filter_text= df['Quantity Ordered'] != 'Quantity Ordered'

df=df[filter_text]

df['Quantity Ordered'].unique()

array(['2', '1', '3', '5', '4', '7', '6', '8', '9'], dtype=object)

In [50]:
#convert the quantity ordered and price each data type
df['Quantity Ordered']=df['Quantity Ordered'].astype('float')
df['Price Each'] = df['Price Each'].astype('float')

In [51]:
df['Order Date']=pd.to_datetime(df['Order Date'])
df.dtypes

Order ID                    object
Product                     object
Quantity Ordered           float64
Price Each                 float64
Order Date          datetime64[ns]
Purchase Address            object
dtype: object

In [52]:
#create the Year, Month , Hour , Minute columns
df['Year']= df['Order Date'].dt.year
df['Month']= df['Order Date'].dt.month
df['Hour']= df['Order Date'].dt.hour
df['Minute']= df['Order Date'].dt.minute

In [53]:
#create the revenue column
df['Revenue']=df['Price Each'] * df['Quantity Ordered']

m='India, vs, SriLanka'
m.split(',')[1]

' vs'

In [54]:
city=[]
for i in df['Purchase Address']:
    c=i.split(',')[1]
    city.append(c)

df['City']=city
df

Order ID                     Product  Quantity Ordered  Price Each  \
0       176558        USB-C Charging Cable               2.0       11.95   
2       176559  Bose SoundSport Headphones               1.0       99.99   
3       176560                Google Phone               1.0      600.00   
4       176560            Wired Headphones               1.0       11.99   
5       176561            Wired Headphones               1.0       11.99   
...        ...                         ...               ...         ...   
11681   259353      AAA Batteries (4-pack)               3.0        2.99   
11682   259354                      iPhone               1.0      700.00   
11683   259355                      iPhone               1.0      700.00   
11684   259356      34in Ultrawide Monitor               1.0      379.99   
11685   259357        USB-C Charging Cable               1.0       11.95   

               Order Date                         Purchase Address  Year  \
0     2019-04-19 08:46:00             917 1st St, Dallas, TX 75001  2019   
2     2019-04-07 22:30:00        682 Chestnut St, Boston, MA 02215  2019   
3     2019-04-12 14:38:00     669 Spruce St, Los Angeles, CA 90001  2019   
4     2019-04-12 14:38:00     669 Spruce St, Los Angeles, CA 90001  2019   
5     2019-04-30 09:27:00        333 8th St, Los Angeles, CA 90001  2019   
...                   ...                                      ...   ...   
11681 2019-09-17 20:56:00   840 Highland St, Los Angeles, CA 90001  2019   
11682 2019-09-01 16:00:00  216 Dogwood St, San Francisco, CA 94016  2019   
11683 2019-09-23 07:39:00     220 12th St, San Francisco, CA 94016  2019   
11684 2019-09-19 17:30:00   511 Forest St, San Francisco, CA 94016  2019   
11685 2019-09-30 00:18:00   250 Meadow St, San Francisco, CA 94016  2019   

       Month  Hour  Minute  Revenue            City  
0          4     8      46    23.90          Dallas  
2          4    22      30    99.99          Boston  
3          4    14      38   600.00     Los Angeles  
4          4    14      38    11.99     Los Angeles  
5          4     9      27    11.99     Los Angeles  
...      ...   ...     ...      ...             ...  
11681      9    20      56     8.97     Los Angeles  
11682      9    16       0   700.00   San Francisco  
11683      9     7      39   700.00   San Francisco  
11684      9    17      30   379.99   San Francisco  
11685      9     0      18    11.95   San Francisco  

[185950 rows x 12 columns]

In [55]:
df.drop('Order Date',axis=1,inplace=True)

In [56]:
df

Order ID                     Product  Quantity Ordered  Price Each  \
0       176558        USB-C Charging Cable               2.0       11.95   
2       176559  Bose SoundSport Headphones               1.0       99.99   
3       176560                Google Phone               1.0      600.00   
4       176560            Wired Headphones               1.0       11.99   
5       176561            Wired Headphones               1.0       11.99   
...        ...                         ...               ...         ...   
11681   259353      AAA Batteries (4-pack)               3.0        2.99   
11682   259354                      iPhone               1.0      700.00   
11683   259355                      iPhone               1.0      700.00   
11684   259356      34in Ultrawide Monitor               1.0      379.99   
11685   259357        USB-C Charging Cable               1.0       11.95   

                              Purchase Address  Year  Month  Hour  Minute  \
0                 917 1st St, Dallas, TX 75001  2019      4     8      46   
2            682 Chestnut St, Boston, MA 02215  2019      4    22      30   
3         669 Spruce St, Los Angeles, CA 90001  2019      4    14      38   
4         669 Spruce St, Los Angeles, CA 90001  2019      4    14      38   
5            333 8th St, Los Angeles, CA 90001  2019      4     9      27   
...                                        ...   ...    ...   ...     ...   
11681   840 Highland St, Los Angeles, CA 90001  2019      9    20      56   
11682  216 Dogwood St, San Francisco, CA 94016  2019      9    16       0   
11683     220 12th St, San Francisco, CA 94016  2019      9     7      39   
11684   511 Forest St, San Francisco, CA 94016  2019      9    17      30   
11685   250 Meadow St, San Francisco, CA 94016  2019      9     0      18   

       Revenue            City  
0        23.90          Dallas  
2        99.99          Boston  
3       600.00     Los Angeles  
4        11.99     Los Angeles  
5        11.99     Los Angeles  
...        ...             ...  
11681     8.97     Los Angeles  
11682   700.00   San Francisco  
11683   700.00   San Francisco  
11684   379.99   San Francisco  
11685    11.95   San Francisco  

[185950 rows x 11 columns]

In [57]:
df.drop('Purchase Address',axis=1,inplace=True)

In [58]:
df

Order ID                     Product  Quantity Ordered  Price Each  \
0       176558        USB-C Charging Cable               2.0       11.95   
2       176559  Bose SoundSport Headphones               1.0       99.99   
3       176560                Google Phone               1.0      600.00   
4       176560            Wired Headphones               1.0       11.99   
5       176561            Wired Headphones               1.0       11.99   
...        ...                         ...               ...         ...   
11681   259353      AAA Batteries (4-pack)               3.0        2.99   
11682   259354                      iPhone               1.0      700.00   
11683   259355                      iPhone               1.0      700.00   
11684   259356      34in Ultrawide Monitor               1.0      379.99   
11685   259357        USB-C Charging Cable               1.0       11.95   

       Year  Month  Hour  Minute  Revenue            City  
0      2019      4     8      46    23.90          Dallas  
2      2019      4    22      30    99.99          Boston  
3      2019      4    14      38   600.00     Los Angeles  
4      2019      4    14      38    11.99     Los Angeles  
5      2019      4     9      27    11.99     Los Angeles  
...     ...    ...   ...     ...      ...             ...  
11681  2019      9    20      56     8.97     Los Angeles  
11682  2019      9    16       0   700.00   San Francisco  
11683  2019      9     7      39   700.00   San Francisco  
11684  2019      9    17      30   379.99   San Francisco  
11685  2019      9     0      18    11.95   San Francisco  

[185950 rows x 10 columns]

#### Questions
- What is the total number of sales?
- What is the average sales per month?
- What was the best month for sales? How much was earned that month?
- What city had the highest number of sales?
- What product sold the most? Why it sold the most?
- What products are most often sold together?
- What time should we display advertisement to maximize likelihood of customer's buying product?

In [61]:
#q1
year_sales=df.groupby("Year").agg({'Revenue':'sum'})
print(year_sales)

          Revenue
Year             
2019  34483365.68
2020      8670.29


In [63]:
#q2
df=df[df.Year!=2020]
df

Order ID                     Product  Quantity Ordered  Price Each  \
0       176558        USB-C Charging Cable               2.0       11.95   
2       176559  Bose SoundSport Headphones               1.0       99.99   
3       176560                Google Phone               1.0      600.00   
4       176560            Wired Headphones               1.0       11.99   
5       176561            Wired Headphones               1.0       11.99   
...        ...                         ...               ...         ...   
11681   259353      AAA Batteries (4-pack)               3.0        2.99   
11682   259354                      iPhone               1.0      700.00   
11683   259355                      iPhone               1.0      700.00   
11684   259356      34in Ultrawide Monitor               1.0      379.99   
11685   259357        USB-C Charging Cable               1.0       11.95   

       Year  Month  Hour  Minute  Revenue            City  
0      2019      4     8      46    23.90          Dallas  
2      2019      4    22      30    99.99          Boston  
3      2019      4    14      38   600.00     Los Angeles  
4      2019      4    14      38    11.99     Los Angeles  
5      2019      4     9      27    11.99     Los Angeles  
...     ...    ...   ...     ...      ...             ...  
11681  2019      9    20      56     8.97     Los Angeles  
11682  2019      9    16       0   700.00   San Francisco  
11683  2019      9     7      39   700.00   San Francisco  
11684  2019      9    17      30   379.99   San Francisco  
11685  2019      9     0      18    11.95   San Francisco  

[185916 rows x 10 columns]

In [75]:
df.groupby("Month").agg({'Revenue':'mean'}).sort_values('Month',ascending=False)

Revenue
Month            
12     184.655913
11     182.074956
10     184.238580
9      180.497387
8      187.648849
7      185.249826
6      190.187565
5      190.305852
4      185.495390
3      185.250471
2      183.884962
1      187.450795

In 2019 ,May was the highest average revenue with USD 190.305852 whereas September has the lowest average Revenue with USD 180.4974

In [83]:
#q4
df.groupby("City").agg({'Revenue':'sum'}).sort_values('Revenue',ascending=False)

Revenue
City                      
 San Francisco  8259719.03
 Los Angeles    5451570.81
 New York City  4662976.81
 Boston         3660315.12
 Atlanta        2795498.58
 Dallas         2766260.45
 Seattle        2747605.49
 Portland       2320337.62
 Austin         1819081.77

In [88]:
#q7
df.groupby('Hour').agg({'Revenue':'sum'}).sort_values('Revenue',ascending=False)

Revenue
Hour            
19    2412938.54
12    2316821.34
11    2300610.24
20    2281716.24
18    2219348.30
13    2155389.80
17    2129361.61
14    2083672.73
21    2042000.86
10    1944286.77
15    1941549.60
16    1904601.31
9     1639030.58
22    1607549.21
8     1192348.97
23    1179304.44
7      744854.12
0      711060.59
1      458561.97
6      448113.00
2      232586.50
5      230564.88
4      162396.09
3      144697.99

In [91]:
#5
product_sales=df.groupby('Product').agg({'Quantity Ordered':'sum','Price Each':'mean','Revenue':'sum'}).sort_values(by='Revenue',ascending=False)
product_sales

Quantity Ordered  Price Each     Revenue
Product                                                             
Macbook Pro Laptop                    4727.0     1700.00  8035900.00
iPhone                                6847.0      700.00  4792900.00
ThinkPad Laptop                       4128.0      999.99  4127958.72
Google Phone                          5531.0      600.00  3318600.00
27in 4K Gaming Monitor                6243.0      389.99  2434707.57
34in Ultrawide Monitor                6199.0      379.99  2355558.01
Apple Airpods Headphones             15657.0      150.00  2348550.00
Flatscreen TV                         4818.0      300.00  1445400.00
Bose SoundSport Headphones           13454.0       99.99  1345265.46
27in FHD Monitor                      7547.0      149.99  1131974.53
Vareebadd Phone                       2067.0      400.00   826800.00
20in Monitor                          4126.0      109.99   453818.74
LG Washing Machine                     666.0      600.00   399600.00
LG Dryer                               646.0      600.00   387600.00
Lightning Charging Cable             23211.0       14.95   347004.45
USB-C Charging Cable                 23971.0       11.95   286453.45
Wired Headphones                     20553.0       11.99   246430.47
AA Batteries (4-pack)                27635.0        3.84   106118.40
AAA Batteries (4-pack)               31012.0        2.99    92725.88

From the graph above we can led that The most sold product was AAA Batteries (4-pack) (31.012k) Vie aiso can tell that AA Batteries (4-pack). USB-C Charging Cable, Lightning Charging Cable, and Wired Headphones sold more than other products. These products probably sold the most because it is an dem that is needed often and is relatively cheap